### PR TITLE
add callback before main loop

### PR DESCRIPTION
**Pull request by d-monnet**

### CODE REVIEW

File: src/MultiPrecisionR2.jl [hunk 1]
```diff
@@ -372,6 +372,8 @@ function SolverCore.solve!(
     )
   end
 
+  callback()
```
Comment (farhadrclass) [inline src/MultiPrecisionR2.jl]: I suggest something like
callback = (args...) -> nothing,

so it is more general with inputs 
We need it in the argument of function

Comment (farhadrclass) [reply src/MultiPrecisionR2.jl]: We need to have something like this callback(nlp, solver, stats)
so we send stats back to the callback and nlp , solver is not important now but in future would be useful


we also need to call cb before the loop and in the loop

Comment (farhadrclass) [reply src/MultiPrecisionR2.jl]: I do not think we need args.. as input of the solver, they are splat operator https://docs.julialang.org/en/v1/manual/functions/#Varargs-Functions

Comment (d-monnet) [reply src/MultiPrecisionR2.jl]: Ok, I just commit the modifications, I'll let you approve the PR if it's all good for you.

File: src/MultiPrecisionR2.jl [hunk 2]
```diff
@@ -371,9 +371,9 @@ function SolverCore.solve!(
       ),
     )
   end
-
-  callback()
-
+  
+  callback(solver,MPnlp,stats)
```
Comment (farhadrclass) [inline src/MultiPrecisionR2.jl]: to be matching all other solver I suggest adding 
callback(MPnlp,solver,stats)

**Review by farhadrclass — COMMENTED** [review]
We also need a unit test ,
Somthing simple for now,
For R2 we did 

```julia
using ADNLPModels, JSOSolvers, LinearAlgebra, Logging #, Plots
@testset "Test callback" begin
  f(x) = (x[1] - 1)^2 + 4 * (x[2] - x[1]^2)^2
  nlp = ADNLPModel(f, [-1.2; 1.0])
  X = [nlp.meta.x0[1]]
  Y = [nlp.meta.x0[2]]
  function cb(nlp, solver, stats)
    x = solver.x
    push!(X, x[1])
    push!(Y, x[2])
    if stats.iter == 8
      stats.status = :user
    end
  end
  stats = with_logger(NullLogger()) do
    R2(nlp, callback = cb)
  end
  @test stats.iter == 8

end


@testset "Testing Solver Values" begin
  f(x) = (x[1] - 1)^2 + 4 * (x[2] - x[1]^2)^2
  nlp = ADNLPModel(f, [-1.2; 1.0])
  function cb(nlp, solver, stats)
    if stats.iter == 4
      @test solver.σ > 0.0
      stats.status = :user
    end
  end
  stats = with_logger(NullLogger()) do
    R2(nlp, callback = cb)
  end
end
```

you need  to modify it a bit and test if the callback works

File: src/MultiPrecisionR2.jl (hunk 2 above)
Comment (d-monnet) [reply src/MultiPrecisionR2.jl]: I'm not sure why you need `X` and `Y` for in the above test, I removed them.

Comment (farhadrclass) [reply src/MultiPrecisionR2.jl]: in this case you do not need it

**Comment by farhadrclass** [comment]
did it pass the test?
if so, please merge it

**Pull request merged**